recovery/elliptics_recovery/types/dc.py

```diff
@@ -234,18 +265,105 @@ def contains(addresses_with_backends, address, backend_id):
+    session.trace_id = ctx.trace_id
+    session.exceptions_policy = elliptics.exceptions_policy.no_exceptions
+    session.set_filter(elliptics.filters.all_final)
+    session.ioflags |= elliptics.io_flags.cas_timestamp
+    session.timestamp = ctx.prepare_timeout
+
+    stats = ctx.stats['recover']
+    stats_cmd = ctx.stats['commands']
+
+    for r in results:
+        with open(r.filename, 'ab') as f:
+            for _, batch in groupby(enumerate(load_key_data(r.uncommitted_filename)),
+                                    key=lambda x: x[0] / ctx.batch_size):
+                batch = [item[1] for item in batch]
+                tasks = []
+                for key, key_infos in batch:
+                    for info in key_infos:
+                        if info.flags & elliptics.record_flags.uncommitted:
+                            tasks.append((key, info.group_id, info.size))
+
+                statuses = {} # (key, group_id) -> status
+                for attempt in range(ctx.attempts):
+                    if not tasks:
+                        break
+
+                    batch_sizes = defaultdict(int) # group_id -> batch_size
+                    for _, group_id, key_size in tasks:
+                        batch_sizes[group_id] += key_size
+
+                    timeouts = {group_id: max(60, batch_size / ctx.data_flow_rate)
+                                for group_id, batch_size in batch_sizes.iteritems()}
+
+                    responses = []
+                    for key, group_id, _ in tasks:
+                        session.groups = [group_id]
+                        session.timeout = timeouts[group_id]
+                        responses.append(session.remove(key))
+
+                    failed_tasks = []
+                    for i, r in enumerate(responses):
+                        key, group_id, _ = tasks[i]
+                        status = r.get()[0].status
+                        log.info('Removed uncommitted key: %s, group: %s, status: %s, attempts: %s/%s',
+                                 key, group_id, status, attempt, ctx.attempts)
+                        statuses[(key, group_id)] = status
+
+                        if status == 0:
+                            stats.counter('removed_corrupted_keys', 1)
+                        else:
+                            stats_cmd.counter('remove.{0}'.format(status), 1)
+
+                        if status not in (0, -errno.ENOENT, -errno.EBADFD):
+                            failed_tasks.append(tasks[i])
+                    tasks = failed_tasks
+
+                for key, key_infos in batch:
+                    infos = []
+                    for info in key_infos:
+                        if info.flags & elliptics.record_flags.uncommitted:
+                            status = statuses[(key, info.group_id)]
+                            if status == -errno.EBADFD:
+                                infos = None
+                                break
```

Comment:
It should also update skipped keys statistics if it excludes keys from recovering.